### PR TITLE
fix(lock): honor negative acquire timeout as wait-forever

### LINKED ISSUE #13719: packages/lock: negative timeout should wait forever, adapters treat it as no-wait
`Utopia\Lock\Lock::acquire()` documents its contract as: 0.0 means do not wait, a negative value means wait forever.

`File::acquire()` returns `$this->tryAcquire()` for every `$timeout <= 0.0`, so a negative timeout is a single non-blocking attempt. Under contention a caller following the documented contract gets an immediate `false`, or a `Contention` exception from `withLock()`, instead of waiting. The Redis `Distributed` adapter and the non-coroutine `Mutex` and `Semaphore` paths have the same shape.

Either the adapters honour negative as unbounded, or the interface doc drops the negative case; pick one and cover it with a unit test on the file lock (contention from a second process or a pre-held handle, negative timeout, assert it blocks until release).

Found by review on #13718 while absorbing the package; left out of that PR because absorption does not change library behaviour. The library now lives at `packages/lock`, so the fix lands here and releases with `bin/monorepo release lock`.

### PR DESCRIPTION
**Pull request by flcrom**
## What does this PR do?

`Utopia\Lock\Lock::acquire()` documents its contract as: `0.0` means do not wait, a negative value means wait forever. `File::acquire()`, `Distributed::acquire()`, and the non-coroutine `Mutex`/`Semaphore` paths all treated every `$timeout <= 0.0` as a single non-blocking attempt, so under contention a caller following the documented contract got an immediate `false` (or a `Contention` from `withLock()`) instead of waiting.

Each path now only short-circuits on exactly `0.0` and retries without a deadline when the timeout is negative. Coroutine paths are untouched (`Channel::push()` already waits indefinitely on a negative timeout). Zero and positive timeout behavior is unchanged, including the jittered backoff in `Distributed`.

## Test Plan

Added `FileTest::testNegativeTimeoutWaitsUntilRelease`, which forks a child process that holds the lock and asserts the parent's `acquire(-1.0)` blocks until the child releases. It fails against the previous implementation and passes with this change.

- `phpunit tests/FileTest.php` on PHP 8.3: OK (5 tests, 16 assertions)
- `php -l` clean on all package files
- `phpstan` at the package's level 9 settings: no new findings

The Swoole-backed Mutex/Semaphore tests and the Redis-backed E2E tier need `ext-swoole` and a Redis server, so those run in CI.

## Related PRs and Issues

- Fixes #13719
- Carved out of #13718 (package absorption), which deliberately left library behavior unchanged

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [x] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs? (No API metadata changes)

### CODE REVIEW

File: packages/lock/src/Distributed.php [hunk 1]
```diff
@@ -60,16 +60,18 @@ public function acquire(float $timeout = 0.0): bool
             return true;
         }
 
-        if ($timeout <= 0.0) {
+        if ($timeout === 0.0) {
             return false;
         }
 
-        $deadline = microtime(true) + $timeout;
+        $deadline = $timeout > 0.0 ? microtime(true) + $timeout : null;
```
[the remark this answers is not in the document]
Comment (flcrom) [reply packages/lock/src/Distributed.php:67]: Added behavioral coverage for the other adapters: a Redis-backed E2E contention test for Distributed (each process opens its own Redis connection after the fork), coroutine contention tests for Mutex and Semaphore asserting the waiter blocks until release, and uncontended sync-path negative-acquire tests for both. Held sync-path contention can't be exercised from a single thread without self-deadlock, so it's covered through those instead.

File: packages/lock/tests/FileTest.php [hunk 2]
```diff
@@ -141,4 +141,42 @@ public function testContentionThrownOnTimeout(): void
...
+            $deadline = microtime(true) + 5.0;
+            while (microtime(true) < $deadline && file_get_contents($ready) === '') {
+                usleep(10_000);
+            }
+
+            $parent = new File($this->path);
+            $start = microtime(true);
+            $this->assertTrue($parent->acquire(-1.0), 'Negative timeout must wait until the lock is released');
+            $elapsed = microtime(true) - $start;
+            $this->assertGreaterThanOrEqual(0.2, $elapsed, 'Acquire must have blocked until the holder released');
```
[the remark this answers is not in the document]
Comment (flcrom) [reply packages/lock/tests/FileTest.php]: Reworked this to synchronize explicitly in both directions: the child reports its tryAcquire result through a ready file that the parent asserts, and a started file gates the child's hold, so the assertion no longer depends on scheduling slack. Cleanup also changed: the parent always signals started before a bounded waitpid (SIGKILL as a last resort), and a new failure-path test forces child acquisition failure to prove the test terminates promptly instead of hanging.

File: packages/lock/tests/E2E/DistributedTest.php [hunk 3]
```diff
@@ -333,4 +334,148 @@ public function testLoggerReceivesMessages(): void
...
+            $waiter->release();
+        } finally {
+            file_put_contents($started, '1');
+            $this->reapChild($pid);
+            @unlink($ready);
+            @unlink($started);
+        }
+    }
+
+    public function testChildAcquisitionFailureTerminatesPromptly(): void
```
[the remark this answers is not in the document]
Comment (flcrom) [reply packages/lock/tests/E2E/DistributedTest.php]: Removed both failure-path tests - fair point, they exercised PHPUnit assertion and reaper internals rather than observable lock behavior. Prompt termination is now structural instead: every child operation is bounded (bounded coordination wait, fixed hold, non-blocking or timeout-carrying I/O), so the harness no longer needs tests that inspect its own internals.

File: packages/lock/tests/FileTest.php [hunk 4]
```diff
@@ -141,4 +142,137 @@ public function testContentionThrownOnTimeout(): void
...
+            $result = pcntl_waitpid($pid, $status, WNOHANG);
+
+            if ($result === $pid || $result === -1) {
+                return;
+            }
+
+            usleep(10_000);
+        }
+
+        posix_kill($pid, SIGKILL);
```
[the remark this answers is not in the document]
Comment (flcrom) [reply packages/lock/tests/FileTest.php]: Reworked cleanup to be portable and never blocking. posix_kill is gone - the package only requires php >=8.3, so no signals. Every Redis operation in the fork test is now bounded (1s connect timeout, OPT_READ_TIMEOUT on both connections, guarded acquire and release, with the key TTL as backstop for a lost release), and reaping is a bounded non-blocking poll: a child that outlives the window is left to be reaped at process exit instead of hanging the suite. Also fixed the stale hold-time comment.
